bqskit/runtime/detached.py

```diff
@@ -0,0 +1,571 @@
+            self.listener = None  # type: ignore
+            self.listening_thread.join()
+            self.listening_thread = None  # type: ignore
+            self.logger.debug('Joined listening thread.')
+
+        # Close outgoing thread
+        if self.outgoing_thread is not None:
+            self.outgoing_thread.join()
+            self.outgoing_thread = None  # type: ignore
+            self.logger.debug('Joined outgoing thread.')
+
+    def _handle_disconnect(self, conn: Connection) -> None:
+        """Disconnect a client connection from the runtime."""
+        self.sel.unregister(conn)
+        conn.close()
+        tasks = self.clients.pop(conn)
+        for task_id in tasks:
+            self._handle_cancel_comp_task(task_id)
+        self.logger.info('Unregistered client.')
+
+    def _handle_new_comp_task(
+        self,
+        conn: Connection,
+        task: CompilationTask,
+    ) -> None:
+        """Convert a :class:`CompilationTask` into an internal one."""
+        mailbox_id = self._get_new_mailbox_id()
+        self.tasks[task.task_id] = (mailbox_id, conn)
+        self.mailbox_to_task_dict[mailbox_id] = task.task_id
+        self.mailboxes[mailbox_id] = [None, False]
+        self.logger.info(f'New CompilationTask: {task.task_id}.')
+
+        self.clients[conn].add(task.task_id)
+
+        internal_task = RuntimeTask(
+            (CompilationTask.run, (task,), {}),
+            RuntimeAddress(-1, mailbox_id, 0),
+            mailbox_id,
+            tuple(),
+            task.logging_level,
+            task.max_logging_depth,
+        )
+
+        self._handle_new_task(internal_task)
+
+    def _handle_new_task(self, task: RuntimeTask) -> None:
+        """Schedule a task on a manager."""
+        # Select manager with least tasks
+        min_tasks = max(self.manager_idle_resources)
+        best_id = self.manager_idle_resources.index(min_tasks)
+        self.manager_idle_resources[best_id] -= 1
+
+        # Assign work
+        manager = self.managers[best_id]
+        self.outgoing.append((manager, RuntimeMessage.SUBMIT, task))
+
+    def _handle_new_tasks(self, tasks: list[RuntimeTask]) -> None:
+        """Schedule many tasks between the managers."""
+        assignments: list[list[RuntimeTask]] = [[] for _ in self.managers]
+        for task in tasks:
+            # Select manager
+            min_tasks = max(self.manager_idle_resources)
+            best_id = self.manager_idle_resources.index(min_tasks)
+            self.manager_idle_resources[best_id] -= 1
+            assignments[best_id].append(task)
+
+        # Assign work
+        for i, assignment in enumerate(assignments):
+            if len(assignment) == 0:
+                continue
+
+            elif len(assignment) == 1:
+                m = (self.managers[i], RuntimeMessage.SUBMIT, assignment[0])
+                self.outgoing.append(m)
+
+            else:
+                n = (self.managers[i], RuntimeMessage.SUBMIT_BATCH, assignment)
```

Comment:
The n vs m here is for the type checker. It will complain since assignment[0] and assignment have different types. The other way around is to add a type annotation on m, which would be very verbose.